### PR TITLE
Add minmax sparse indexes when compressing columns with btree indexes

### PR DESCRIPTION
**Pull request by akuzm**
The decision to add a minmax sparse index is made every time when the compressed chunk is created (full decompression followed by compression) based on the currently present indexes on the hypertable. No new chunk compression settings are added.

No action is required on upgrade, but the feature is not enabled on existing chunks. The minmax index will be added when the chunk is fully decompressed and compressed.

No action is required on downgrade, we ignore the unknown metadata columns. They will be removed when the chunk is fully decompressed and compressed.

The potential drawback of this feature is increasing the storage requirements for the compressed chunk table, but it is normally only a few percent of the total compressed data size. It can be disabled with the GUC `timescaledb.auto_sparse_indexes`.

Here's a small example of this feature in action: https://gist.github.com/akuzm/84d4b3b609e3581768173bd21001dfbf
Note that the number of hit buffers is reduced almost 4x.

### CODE REVIEW

File: tsl/src/compression/create.c [hunk 1]
```diff
@@ -158,11 +191,19 @@ build_columndefs(CompressionSettings *settings, Oid src_relid)
 		 * Put the metadata columns before the compressed columns, because they
 		 * are accessed before decompression.
 		 */
-		if (is_orderby)
+		if (is_segmentby)
+		{
+			/* No additional metadata for segmentby. */
+		}
```
Comment (fabriziomello) [inline tsl/src/compression/create.c:197]: Do we really need this?

File: tsl/src/compression/create.c [hunk 2]
```diff
@@ -118,15 +121,45 @@ build_columndefs(CompressionSettings *settings, Oid src_relid)
...
+	ListCell *lc;
+	List *index_oids = RelationGetIndexList(rel);
+	foreach (lc, index_oids)
+	{
+		Oid index_oid = lfirst_oid(lc);
+		Relation index_rel = index_open(index_oid, AccessShareLock);
+		IndexInfo *index_info = BuildIndexInfo(index_rel);
+		index_close(index_rel, NoLock);
+
+		if (index_info->ii_Am != BTREE_AM_OID)
```
Comment (fabriziomello) [inline tsl/src/compression/create.c:139]: Sorry for dumb question but what about partial indexes (aka `index_info->ii_Predicate != NIL`) ??

File: tsl/src/compression/create.c (hunk 1 above)
Comment (akuzm) [reply tsl/src/compression/create.c:197]: The rest of conditions have to be under `!is_segmentby`, so I wrote it this way to decrease indentation :) Looks a little weird I agree, will switch it to the more traditional form.

File: tsl/src/compression/create.c (hunk 2 above)
Comment (akuzm) [reply tsl/src/compression/create.c:139]: Not sure about what's the best approach in this case, but at least even if we add this metadata, the result should still be correct. Maybe we should create minmax all the columns from predicate as well.

**Review by fabriziomello — APPROVED** [review]

File: tsl/src/compression/create.c (hunk 2 above)
Comment (svenklemm) [reply tsl/src/compression/create.c:139]: Why are we tying this to btree?

File: tsl/src/compression/create.h [hunk 3]
```diff
@@ -15,10 +15,12 @@
 #define COMPRESSION_COLUMN_METADATA_COUNT_NAME COMPRESSION_COLUMN_METADATA_PREFIX "count"
 #define COMPRESSION_COLUMN_METADATA_SEQUENCE_NUM_NAME                                              \
 	COMPRESSION_COLUMN_METADATA_PREFIX "sequence_num"
-#define COMPRESSION_COLUMN_METADATA_MIN_COLUMN_NAME "min"
-#define COMPRESSION_COLUMN_METADATA_MAX_COLUMN_NAME "max"
+
 #define COMPRESSION_COLUMN_METADATA_PATTERN_V1 "_ts_meta_%s_%d"
 
+/* Version 2 is indexed by the uncompressed chunk attno, not by the orderby index. */
+#define COMPRESSION_COLUMN_METADATA_PATTERN_V2 "_ts_meta_v2_%s_%d"
```
Comment (svenklemm) [inline tsl/src/compression/create.h:22]: Should we use the column name here instead of thte attno?

Comment (svenklemm) [reply tsl/src/compression/create.h:22]: dump/restore may change attno so i dont think its safe to use in this context

**Review by svenklemm — CHANGES_REQUESTED** [review]
I dont think we can safely use attno as part of column name as that might change during dump/restore.

File: tsl/src/compression/create.h (hunk 3 above)
Comment (akuzm) [reply tsl/src/compression/create.h]: Changed to use the column names.

File: tsl/src/compression/create.c (hunk 2 above)
Comment (akuzm) [reply tsl/src/compression/create.c:179]: > Why are we tying this to btree?

The idea is that btree and sparse minmax index can be applied for similar types of queries.

If we have a gin index on a text column, the sparse minmax won't help the text search queries, so no point in creating it.

If we have a brin index, we have to look at the opclass -- if it's a bloom filter, minmax probably will be not so useful, but if it's a range brin, then minmax works. I haven't implemented this check yet.

**Comment by svenklemm** [comment]
Have you thought about persisting this information in the compression settings in our catalog? We probably want to make this configurable at some point. So even if we dont allow direct modification now, having this information in our catalog would be a step in that direction.

**Comment by akuzm** [comment]
> Have you thought about persisting this information in the compression settings in our catalog? We probably want to make this configurable at some point. So even if we dont allow direct modification now, having this information in our catalog would be a step in that direction.

For current functionality, the custom catalog info won't really help, I think it even complicates the things -- e.g. we have to think about how the implicit sparse index creation based on uncompressed btree is going to interact with the explicitly specified indexes. I'll probably have to implement fully configurable sparse indexes to understand this well. Do you think we should do it upfront, or this PR is usable enough on its own?

**Pull request merged**